src/system.c

```diff
@@ -25,6 +25,22 @@ void emu_update_uart_interrupts(riscv_t *rv)
     plic_update_interrupts(attr->plic);
 }
 
+#define CLEAR_PENDING_SIGNAL_R()                                    \
+    do {                                                            \
+        if (rv->csr_sepc != rv->last_csr_sepc) {                    \
+            need_handle_signal = true;                              \
+            return 0; /* early return and jump to signal handler */ \
+        }                                                           \
+    } while (0)
+
+#define CLEAR_PENDING_SIGNAL_W()                                  \
```

Comment:
Do we just introduce these macros for repeated code? I think the readability is bad to `return` from a function in a macro...

Reply:
Alternatively, I think we can remain `return` in `CLEAR_PENDING_SIGNAL_*`  and mention they might early return using a comment at the call point of `CLEAR_PENDING_SIGNAL_*`?

Reply:
Is there any disadvantage to remove these macros and just put then inline?

```cpp
need_handle_signal = (rv->csr_sepc != rv->last_csr_sepc);
if (need_handle_signal)
    return 0;

...

need_handle_signal = (rv->csr_sepc != rv->last_csr_sepc);
if (need_handle_signal)
    return;
```

Reply:
> Is there any disadvantage to removing these macros and just putting them inline? It can be just a few lines.
> 
> ```c++
> if (need_handle_signal = (rv->csr_sepc != rv->last_csr_sepc))
>     return 0;
> 
> ...
> 
> if (need_handle_signal = (rv->csr_sepc != rv->last_csr_sepc))
>     return;
> ```

Looking these few LoC is hard to understand (maybe for newcomer), thus abstracting them into a macro and comment around the macro make things clearer.

Reply:
How about we can inline this directly?

```
need_handle_signal = (rv->csr_sepc != rv->last_csr_sepc);
if (need_handle_signal)
    return 0;
```

You can replace `need_handle_signal = (rv->csr_sepc != rv->last_csr_sepc)` with an abstract if you think it need. I am just thinking that we should avoid to `return` in a function-like macro as possible, because it is not clear that the function will return there at the first glance.